### ISSUE TITLE
Tileset: Neither Autotile nor Atlas Tile Priority allow "per Tile" priority

### ISSUE THREAD

**Issue by golddotasksquestions** [godotengine/godot#36972]
**Godot version:** 3.2.1.stable

**OS/device including version:** Win64

**Issue description:**
My goal is to have a Tile with 8 Subtiles and every time I paint a tile, a random one out of those 8 is picked. This is what the "Priority" tab in the Tileset is for. 

Both Atlas texture as well as a Autotile have Priority tab. So I assumed a Tile with 8 identical bitmasks, should provide me with 8 randomly chosen Tiles, if every tile is set to the default 1/8 priority. [At least according to GDQuest](https://youtu.be/F6VerW98gEc?t=333), 1/8  should mean that out of 8 painted tiles, the probability for each subtile to appear is 1.

I would expect this to work, however what seems to be the case is that priority only takes effect, if there are multiple tiles for the same bitmask, **and only if the bitmask _rules_ return true.**

As a consequence, despite **Atlas Tiles** having a "Priority" tab, it seems to have absolutely no purpose or effect:
![Tileset_priority_atlas](https://user-images.githubusercontent.com/47016402/76356728-b5dfe100-6316-11ea-80cb-961949bc27ff.png)

**Autotiles** _without_ a bitmask set will also show no effect of priority, just like with Atlas tiles:
![tileset_priority_autotile_nobitmask](https://user-images.githubusercontent.com/47016402/76359531-b3cc5100-631b-11ea-8717-2dc1de8cfd01.gif)
Note the 1/0 there.

If you set the 3x3 bitmask to fill the entire tile, all tiles surrounded by other tiles, will be affected by "priority", but not those that are not surrounded at all sides:
![tileset_priority_autotile_full](https://user-images.githubusercontent.com/47016402/76356909-00615d80-6317-11ea-9bf4-233177597d92.gif)

If you set the 3x3 bitmask only in the center of each tile, the opposite is the case, only tiles not surrounded by any other tiles will be affected by the priority setting:
![tileset_priority_autotile_center](https://user-images.githubusercontent.com/47016402/76360279-f3476d00-631c-11ea-98dc-2b5cef55defd.gif)

The desired result however, would be to allow the user control over which tiles are set according to priority and which are not. The way bitmaps are currently set up, seem to make the priority feature either not working or useless.

See it in action:
![tileset_priority](https://user-images.githubusercontent.com/47016402/76360465-56d19a80-631d-11ea-9d52-7147bf63a1f0.gif)

**Minimal reproduction project:**
[Tileset_Priority_issue.zip](https://github.com/godotengine/godot/files/4314839/Tileset_Priority_issue.zip)

PS: If you wonder why Atlas tiles don't show up in the right color: https://github.com/godotengine/godot/issues/36964

**Comment by zzz-assault** [comment]
Just stumbled about this issue, as I wanted to optimize my proc gen programm with ATLAS instead of AUTO_TILE. And I can confirm, that the priority system of ATLAS has no effect at all, it will always pick 1st tile / icon tile.

Regarding AUTO_TILE, here the priority system is working for my needs, you just have to use the WILDCARD function of the bitmask settings (i. e. set all 9 bits to WILDCARD).

**Comment by golddotasksquestions** [comment]
>  you just have to use the WILDCARD function of the bitmask settings (i. e. set all 9 bits to WILDCARD

How? I can't find anything about a wildcard in the Tileset docs, be it the tutorial or the API, be it 3.2 or latest.

**Comment by zzz-assault** [comment]
@golddotasksquestions Wildcard Bits can be set by using "Shift + LMB" in the autotile editor. They are marked with a blue checker board pattern.

You can find more details in the PR #27565
 And i'm sure that there should ecist a youtube tutorial, just search for wildcard.

**Comment by golddotasksquestions** [comment]
Thanks, but what does shift-clicking do? What does "adding a wildcard" mean in the context of Tileset bitmasks?
There is no tutorial on Youtube and no documentation. I don't understand C++ code.

**Comment by zzz-assault** [comment]
I can't provide any documentation or tutorial, but bitmask means, that it's trying to match correct tileimages based on corresponding tiles. The WILDCARD simply ignores this bit (it's irrelevant if it's hit or not hit the same tile), so it's possible to use less images for all combinations if not needed.

For your use case, set ALL 9 (means also Center) bits of the 3x3 minimal to WILDCARD, then it should work.

**Comment by golddotasksquestions** [comment]
Thanks for the explanation!

> For your use case, set ALL 9 (means also center) bits of the 3x3 minimal to WILDCARD, then it should work.

Unfortunately for my usecase it does not work because tiles that have already been set will change if there is an adjacent tile set.
I suppose this is what the Atlas priority would suitable for, but with Atlas tiles priority exists but does not work at all.

**Comment by zzz-assault** [comment]
Lucky you, while looking at the ATLAS Tilemap mode, I've found the missing part to get it running within the editor -> you need to enable priority in the TILEMAP for the ATLAS tile, then it should work as you expect.

![enable-priority](https://user-images.githubusercontent.com/47723585/81121101-40674a00-8f2e-11ea-8519-193e95657c29.png)

**BUT:** I haven't found a way to get it working for "proc gen" content, I assume there is a deadlock / fixed connection between the editor setting (mentioned above) and the priority system. Need to check it more in detail.

**Comment by zzz-assault** [comment]
@golddotasksquestions With the "enable priority" flag and also the PR created (hopefully gets merged soon), all mentioned issues should be covered. Or still anything not working?

**Comment by golddotasksquestions** [comment]
I have no idea how I could have missed that. 
Maybe I just have only looked at that spot when I had an Autotile selected. In that case it says "Disable Autotile".
Anyways, thanks a lot @zzz-assault !

> With the "enable priority" flag and also the PR created (hopefully gets merged soon), all mentioned issues should be covered. Or still anything not working?
Yes!
Being able to use that feature when generating Tilemaps would be very much desired! 

EDIT:
After testing for a while with Atlastiles Priority enabled, I noticed the surrounding tiles changing again, but only when erasing the second time! :(
I guess this means it's not completely fixed ...

![erasing_atlastiles_neighbors_issue](https://user-images.githubusercontent.com/47016402/81444801-0db09200-9178-11ea-9cf8-acdd7b51224f.gif)

**Comment by zzz-assault** [comment]
Yes, I know this behavoir. It is also stated in the PR (as additional remark).
But the fix for it is not yet included in the PR, as this could have negative impact on parallel used AUTO-TILES -> therefore I put it to discussion in the PR.

> Additional Remark: (not done change)
> I would also delete the following code in "TileMapEditor::_set_cell", as all ATLAS tile interactions are done based on the actual "cell", only in this case (which is deletion of tiles in the Tilemap Editor) it's updating the "bitmask_area" (3x3 field of cells), which causes also to reset all surrounding tiles. I don't think that this was the intention of the implementation and could also confuse users (only drawback I can see = it could have an impact on AUTO-TILES directly surrounding the deleted ATLAS tile). Remark: no need to change to "update_cell_bitmask" as the cell is set to empty in this case (-1).
> ```
> 	} else {
> 		node->update_bitmask_area(Point2(p_pos));
> 	}
> ```

**Comment by guilhermefelipecgs** [comment]
> After testing for a while with Atlastiles Priority enabled, I noticed the surrounding tiles changing again, but **only when erasing the _second_ time**! :(
> I guess this means it's not completely fixed ...

This happens because we need to guarantee that surrounding tiles that have bitmask (if is autotile for example) will be updated correctly. The way the tilemaps are implemented, we cannot distinguish what are the tile type was used (atlas, auto or single) for each cell.

**Comment by guilhermefelipecgs** [comment]
And if you're using a random tile (based on priority), I don't see a problem updating it.

**Comment by AllenKll** [comment]
There's a lot of discussion about Atlas, but regarding autotile, is there a fix for this planned too?

**Comment by guilhermefelipecgs** [comment]
@AllenKll Autotile already works, but you need to use a 3x3 bitmask and setup wildcards to get the same result.

Like this:
![Screenshot from 2020-05-23 14-06-22](https://user-images.githubusercontent.com/1387165/82736248-9eea4180-9cfe-11ea-8f13-8d3ea774aa4b.png)

**Comment by golddotasksquestions** [comment]
@AllenKll @guilhermefelipecgs 
This Autotile set up (center bit, plus wildcard all around) has the same issue that exists when erasing the Atlas tiles: the neighboring, already set tiles get changed too. Only difference is with Atlas tiles it only happens when erasing, with Autotiles it happens when erasing _or_ drawing tiles next to existing tiles. I suppose that's what AllenKll meant.

**Comment by guilhermefelipecgs** [comment]
@golddotasksquestions For autotile this is expected, it necessary to recalculate bitmask/priority every time you draw/erase.

**Comment by golddotasksquestions** [comment]
@guilhermefelipecgs I don't understand why you say this is expected. If I put the wildcard all around, I would expect no change, regardless what happens around.

**Comment by golddotasksquestions** [comment]
I just tested this issue in Godot 3.4 and it is still persistent.
```

func _ready():
	for x in 10:
		for y in 5:
			$TileMap.set_cell(x, y, 0)
```

When you try to set an Atlas Tile in code, it won't take the Priority setting into account at all. This make Atlas Tiles useless for procedural generation. Which is bad, because the only alternative (Autotiles), will also change the neighboring tiles.
